### PR TITLE
Streamdeck + support

### PR DESCRIPTION
**Pull request by karstlok**
Wrote support for the StreamDeck +. Needed to change some really core functionality to have rotary-pushes and rotary-turns passed to callback functions, and to still support older implementations.
It's working for me with a StreamDeck+, but don't have another deck here to test.

Todo: LCD Touch events, Testing with other hardware

Example for new callbacks:
  
        def rotary_change(deck, rotary, rotary_state):
            print("rotary pushed: " + str(rotary) + " state: " + str(rotary_state))
  
        def rotary_turned(values):
            print("rotary turned: " + str(values))

        deck.set_rotarypush_callback(rotary_change)
        deck.set_rotaryturn_callback(rotary_turned)

Example to set touchlcd:
        
        img = Image.new('RGB', (800, 100), 'yellow')
        img_byte_arr = io.BytesIO()
        img.save(img_byte_arr, format='JPEG')
        img_byte_arr = img_byte_arr.getvalue()
        deck.set_lcd_image(0, 0, 800, 100, img_byte_arr)

### CODE REVIEW

File: src/example_plus.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+from StreamDeck.ImageHelpers import PILHelper
+from StreamDeck.Transport.Transport import TransportError
+
+# Folder location of image assets used by this example.
+ASSETS_PATH = os.path.join(os.path.dirname(__file__), "Assets")
+
+
+# image for idle state
+img = Image.new('RGB', (120, 120), color='black')
+released_icon = Image.open(os.path.join(ASSETS_PATH, 'released.png')).resize((80,80))
```
Comment (jpoppe) [inline src/example_plus.py:29]: released.png -> Released.png

File: src/example_plus.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+# callback when rotaries are pressed or released
+def rotary_change(deck, rotary, rotary_state):
+    print("rotary pushed: " + str(rotary) + " state: " + str(rotary_state))
+    if rotary == 3 and rotary_state:
+        deck.reset()
+        deck.close()
+    else:
+        # build an image for the touch lcd
+        img = Image.new('RGB', (800, 100), 'black')
+        icon = Image.open(os.path.join(ASSETS_PATH, 'exit.png')).resize((80,80))
```
Comment (jpoppe) [inline src/example_plus.py:64]: exit.png -> Exit.png

File: src/example_plus.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        deck.set_brightness(100)
+
+        
+        for key in range(0, deck.KEY_COUNT):
+            deck.set_key_image(key, img_released_bytes)
+
+
+        # build an image for the touch lcd
+        img = Image.new('RGB', (800, 100), 'black')
+        icon = Image.open(os.path.join(ASSETS_PATH, 'exit.png')).resize((80,80))
```
Comment (jpoppe) [inline src/example_plus.py:115]: exit.png -> Exit.png

**Review by jpoppe — COMMENTED** [review]
Some `file systems` are `case sensitive` (like Linux ext4 for example) so file names should match capitals, I've got the `example` working after those suggested changes 🎉 

Thanks for your work, can't wait for touch screen support, to busy here ATM 😇

**Comment by karstlok** [comment]
Hi! Thank you for reviewing. I'll make sure to the fix those case-issues in the next commit. 
Working on touch support as well, maybe I'll have time for that somewhere this weekend, but it sure is coming :)

**Comment by karstlok** [comment]
So a few commits later and there is an example, stuff is more or less organized the way it's supposed to and: Touch is working as well. So I'd say the + is fully supported!

**Comment by karstlok** [comment]
Just also tested this against a v1 and all working as expected.

**Comment by abcminiuser** [comment]
Will review as soon as I can, thank you and well done @karstlok!

**Comment by abcminiuser** [comment]
Okiedokie; sorry about the delay, we've had some family medical issues to work through.

I've just done a quick pull, minor cleanup and merge against master in https://github.com/abcminiuser/python-elgato-streamdeck/tree/streamdeckplus. Looking over the changes:

1) I think it might be better to rename `_read_key_states()` to `_read_control_states()` (fine since it's a private/internal method) and have that return a dict of changes, e.g. `{ 'keys': [...], 'rotary_push': [...] }`. That pushes the product specific report parsing into the product specific implementation files, and out of the base StreamDeck.py file.

2) It might be time for me to add in some capability flags and APIs, and deprecate the `is_visual()` and other methods that are in the existing API. Having a single method or more focused properties probably makes more sense to discover the device's actual capabilities.

3) Perhaps `TOUCH_EVENT_SHORT` and the other constants should be part of a proper nested Python enum (`https://docs.python.org/3/library/enum.html`).

Thoughts on the above? I wasn't too sure where Elgato/Corsair were going with the StreamDeck devices when they first launched, but it's getting more obvious to me as new products come out that my API isn't going to scale well as they start making the products more general purpose and varied in physical capabilities.

**Comment by basnijholt** [comment]
@karstlok, it would be great to have support for this. 

Some folks want to run https://github.com/basnijholt/home-assistant-streamdeck-yaml with a StreamDeck + 😄

**Comment by zach2825** [comment]
Howdy peeps, when is the plan to merge this? It would be nice to have a functioning, expensive device and not hold it up for features :)

**Comment by benedikt-bartscher** [comment]
Hey guys, thanks for your work on Streamdeck + Support. I just tested this PR with my new Streamdeck + and so far it works perfectly fine.

**Comment by yterrettaz** [comment]
Hello, thank you for this PR. Also tested today and it works fine with my new Stream Deck +
- Button OK
- LCD OK
- Rotaries OK

**Comment by brimston3** [comment]
Honestly, I'd prefer that rotaryturn events and the key/rotarypush events had nearly the same signature, eg. rotaryturn_callback(self, knob_number, distance) signature. For the touch events, I would hope there's never more than one touch surface. That being said, I'd rather things get merged rather than hold them up.

**Comment by brimston3** [comment]
I added a PR karstlok/python-elgato-streamdeck#1 for the items I mentioned, additional context for rotary and touch callbacks. It closes #122.

**Comment by abcminiuser** [comment]
Thanks! I meant to get to this sooner, but I've merged a slightly tweaked version of this (with extra changes from another contributor) in 56f91960297f753acdaeb79fd460c4326f7c4840.

Unfortunately the StreamDeck + is damned expensive in Australia (and I can't justify the expense for a hobby project) and Corsair told me to shove it when I asked for even temporary access to a unit, so I'm trusting on you and the rest of the community to verify that it works before I tag off the next version for PyPi.

Can you please grab the latest master and verify the example script works on the StreamDeck +?

**Comment by jni** [comment]
@abcminiuser Just saw that you're in Melbourne! I'm in Geelong and I have been monitoring this PR to wait to buy one! If you're reasonably sure you can get all the features working (or help me get all the features working 😅) I am more than happy to get one and bring it over. 😃 I'm reasonably sure I can get it reimbursed by Monash, too. 😂 If you're interested you can email me at juan.nunez-iglesias@monash.edu and we can set something up!

**Comment by jni** [comment]
Looks like you got plenty of tests on #127/#129, but anyway, +1 here. (Rode my bike 20km to JB, got swooped by a magpie 🦅, but it friggin works! 😂 😃)

**Comment by abcminiuser** [comment]
Thanks for testing @jni ! I'll make a new release with support integrated for Pypi soon.

If it's confirmed working now I suspect it'll be a hard ask for reimbursement from your uni for more hardware, but as long as the community can check future changes we can muddle along.

I am indeed in Melbourne! Funny how these things end up, this isn't the first time I've been serendipitously close to someone that uses one of my projects.

**Pull request closed without merging**